Tools/clinic/clinic.py

```diff
@@ -548,43 +556,49 @@ def declare_parser(*, hasformat=False):
     else:
         fname = '.fname = "{name}",'
         format_ = ''
-    declarations = """
-        #define NUM_KEYWORDS {num_keywords}
-        #if NUM_KEYWORDS == 0
-
-        #  if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
-        #    define KWTUPLE (PyObject *)&_Py_SINGLETON(tuple_empty)
-        #  else
-        #    define KWTUPLE NULL
-        #  endif
-
-        #else  // NUM_KEYWORDS != 0
-        #  if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
-
-        static struct {{
-            PyGC_Head _this_is_not_used;
-            PyObject_VAR_HEAD
-            PyObject *ob_item[NUM_KEYWORDS];
-        }} _kwtuple = {{
-            .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, NUM_KEYWORDS)
-            .ob_item = {{ {keywords_py} }},
-        }};
-        #  define KWTUPLE (&_kwtuple.ob_base.ob_base)
-
-        #  else  // !Py_BUILD_CORE
-        #    define KWTUPLE NULL
-        #  endif  // !Py_BUILD_CORE
-        #endif  // NUM_KEYWORDS != 0
-        #undef NUM_KEYWORDS
-
-        static const char * const _keywords[] = {{{keywords_c} NULL}};
-        static _PyArg_Parser _parser = {{
-            .keywords = _keywords,
-            %s
-            .kwtuple = KWTUPLE,
-        }};
-        #undef KWTUPLE
-        """ % (format_ or fname)
+    num_keywords = len(keywords(f))
+    if num_keywords == 0:
+        declarations = """
+            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
+            #  define KWTUPLE (PyObject *)&_Py_SINGLETON(tuple_empty)
+            #else
+            #  define KWTUPLE NULL
+            #endif
+
+            static const char * const _keywords[] = {{{keywords_c} NULL}};
+            static _PyArg_Parser _parser = {{
+                .keywords = _keywords,
+                %s
+                .kwtuple = KWTUPLE,
+            }};
+            #undef KWTUPLE
+            """ % (format_ or fname)
+    else:
+        declarations = """
+            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
+
+            static struct {{
+                PyGC_Head _this_is_not_used;
+                PyObject_VAR_HEAD
+                PyObject *ob_item[%d];
+            }} _kwtuple = {{
+                .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, %d)
+                .ob_item = {{ {keywords_py} }},
+            }};
+            #define KWTUPLE (&_kwtuple.ob_base.ob_base)
+
+            #else  // !Py_BUILD_CORE
+            #  define KWTUPLE NULL
+            #endif  // !Py_BUILD_CORE
+
+            static const char * const _keywords[] = {{{keywords_c} NULL}};
+            static _PyArg_Parser _parser = {{
+                .keywords = _keywords,
+                %s
+                .kwtuple = KWTUPLE,
+            }};
+            #undef KWTUPLE
+            """ % (num_keywords, num_keywords, format_ or fname)
```

Comment:
We could simplify a little.  Something like this:

```suggestion
    if num_keywords == 0:
        declarations = """
            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
            #  define KWTUPLE (PyObject *)&_Py_SINGLETON(tuple_empty)
            #else
            #  define KWTUPLE NULL
            #endif
            """"
    else:
        declarations = """
            #if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)

            static struct {{
                PyGC_Head _this_is_not_used;
                PyObject_VAR_HEAD
                PyObject *ob_item[%d];
            }} _kwtuple = {{
                .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, %d)
                .ob_item = {{ {keywords_py} }},
            }};
            #define KWTUPLE (&_kwtuple.ob_base.ob_base)

            #else  // !Py_BUILD_CORE
            #  define KWTUPLE NULL
            #endif  // !Py_BUILD_CORE
            """ % (num_keywords, num_keywords)

    declarations += """

        static const char * const _keywords[] = {{{keywords_c} NULL}};
        static _PyArg_Parser _parser = {{
            .keywords = _keywords,
            %s
            .kwtuple = KWTUPLE,
        }};
        #undef KWTUPLE
        """ % (format_ or fname)
```

Reply:
Yes, that makes sense.